sunpy/map/maphelper.py

```diff
@@ -0,0 +1,246 @@
+    """
+
+    if not isinstance(coordinate, SkyCoord):
+        raise ValueError("coordinate needs to be a SkyCoord instance.")
+
+    if coordinate.obstime is None:
+        raise ValueError("The coordinate needs an observation time, `obstime`.")
+
+    coordinate = coordinate.transform_to(frames.Helioprojective)
+
+    meta_wcs = _get_wcs_meta(coordinate)
+
+    meta_observer = _get_observer_meta(coordinate)
+
+    meta_instrument = _get_instrument_meta(**kwargs)
+
+    meta_wcs.update(meta_observer)
+    meta_wcs.update(meta_instrument)
+
+    if reference_pixel is None:
+        reference_pixel = u.Quantity([(data.shape[1] + 1)/2.*u.pixel, (data.shape[0] + 1)/2.*u.pixel])
+    if scale is None:
+        scale = u.Quantity([1.0*u.arcsec, 1.0*u.arcsec])
+
+    meta_wcs['crval1'], meta_wcs['crval2'] = coordinate.Tx.to_value(meta_wcs['cunit1']), coordinate.Ty.to_value(meta_wcs['cunit2'])
+
+    meta_wcs['crpix1'], meta_wcs['crpix2'] = reference_pixel[0].to_value(u.pixel), reference_pixel[1].to_value(u.pixel)
+
+    meta_wcs['cdelt1'], meta_wcs['cdelt2'] = scale[0]*meta_wcs['cunit1']/u.pixel, scale[1]*meta_wcs['cunit2']/u.pixel
+
+    meta_dict = MetaDict(meta_wcs)
+
+    for key in kwargs:
+        if key in map_meta_keywords:
+            meta_dict[key] = kwargs[key]
+
+    return meta_dict
+
+
+def _get_wcs_meta(coordinate):
+    """
+    Function to get WCS meta from the SkyCoord using
+    `astropy.wcs.utils.celestial_frame_to_wcs`
+
+    Parameters
+    ----------
+    coordinate : ~`astropy.coordinates.SkyCoord`
+
+    Returns
+    -------
+    `dict`
+        Containing the WCS meta information
+            * ctype1, ctype2
+            * cunit1, cunit2
+            * date_obs
+    """
+
+    coord_meta = {}
+
+    skycoord_wcs = astropy.wcs.utils.celestial_frame_to_wcs(coordinate.frame)
+
+    cunit1, cunit2 = skycoord_wcs.wcs.cunit
+    coord_meta = dict(skycoord_wcs.to_header())
+    coord_meta['cunit1'], coord_meta['cunit2'] = cunit1, cunit2
+
+    return coord_meta
+
+
+def _get_observer_meta(coordinate):
+    """
+    Function to get observer meta from the SkyCoord using
+    `astropy.wcs.utils.celestial_frame_to_wcs`
+
+    Parameters
+    ----------
+    coordinate : ~`astropy.coordinates.SkyCoord`
+
+    Returns
+    -------
+    `dict`
+        Containing the WCS meta information
+            * hgln_obs, hglt_obs
+            * dsun_obs
+            * rsun_obs
+    """
+    coord_meta = {}
+
+    skycoord_wcs = astropy.wcs.utils.celestial_frame_to_wcs(coordinate.frame)
+
+    coord_meta['hgln_obs'], coord_meta['hglt_obs'] = skycoord_wcs.heliographic_observer.lon.to_value(), skycoord_wcs.heliographic_observer.lat.to_value()
+    coord_meta['dsun_obs'] = skycoord_wcs.heliographic_observer.radius.to(u.m).to_value()
+    coord_meta['rsun_ref'] = skycoord_wcs.rsun.to(u.m).to_value()
+    coord_meta['rsun_obs'] = ((skycoord_wcs.rsun/skycoord_wcs.heliographic_observer.radius).decompose() * u.radian).to(u.arcsec).to_value()
```

Comment:
As @wafels pointed out this needs to use the full equation not the small angle approximation.